forgotten-sender/listing-04-fix-2/main.go

```diff
@@ -0,0 +1,72 @@
+// Example program with the goroutine leak fixed. The code cancels the worker
+// on its way out of the process function.
+package main
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"runtime"
+	"time"
+)
+
+func main() {
+
+	// Seed the random number generator so we get different results.
+	// The seed 42 always results in a timeout.
+	// The seed 99 always results in success.
+	// In production we may seed using the current time like
+	// rand.Seed(time.Now().UnixNano())
+	rand.Seed(42)
+
+	fmt.Printf("Number of goroutines: %d\n\n", runtime.NumGoroutine())
+
+	process(context.Background())
+
+	// Sleep long enough to ensure the goroutine has finished.
+	time.Sleep(200 * time.Millisecond)
+
+	fmt.Printf("\nNumber of goroutines: %d\n", runtime.NumGoroutine())
+}
+
+func process(ctx context.Context) {
+
+	// Wrap the parent context in a new context with a 100ms deadline. The
+	// context will be canceled if 100ms passes or if the parent function's
+	// context is canceled for any reason.
+	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+	defer cancel()
+
+	// Make a channel for our goroutine to report its result.
+	ch := make(chan string)
```

Comment:
You still need a buffer of 1

Reply:
Don't go backwards on code we jsut taught

Reply:
I disagree. This part of the code is presented as an alternative solution to using the buffered channel, not a further improvement. An unbuffered channel is a better choice here in terms of both resource usage and correctness.

From the resource usage stance: If the worker knows it is canceled it doesn't **need** the buffer because it can abort.

From the correctness stance: If there **is** buffer space the `select` block may choose to place the value there rather than executing the `<-ctx.Done()` case. We might interpret this that the worker successfully sent the value but really it didn't. See my updated code where I added `fmt` calls to print "Worker completed" and "Worker canceled".

Reply:
You have timing issues and therefore a race. if it is "correct" to do something, then don't teach against it for any reason.

Reply:
I will state again, you just taught, as we do in the classroom, if you dont see a buffer of 1 in a cancellation pattern, you will have problems.